### PR TITLE
`create-release.yml`: Release directly from the `master` branch

### LINKED ISSUE #3732: Remove the `release` branch, and instead release from the `master` branch
Compare the following folders:

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/release/dev
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/master/dev

The master branch currently has many development files that the release branch is missing.

This means that when we changed the `dev/isct_dice_coefficient/CMakeLists.txt` file on `master` in #3710, this caused a [merge conflict](https://github.com/spinalcordtoolbox/spinalcordtoolbox/runs/5469116099?check_suite_focus=true#step:5:48) with our "Create a release" workflow:

```
CONFLICT (modify/delete): dev/isct_dice_coefficient/CMakeLists.txt deleted in HEAD and modified in master.  Version master of dev/isct_dice_coefficient/CMakeLists.txt left in tree.
```

This kind of issue would be mitigated greatly if we removed the `release` branch, and instead released directly from `master`.

### PR DESCRIPTION
**Pull request by joshuacwnewton**
#### GitHub

- [ ] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/geek-tips/contributing#pr-labels-a-href-pr-labels-id-pr-labels-a) to this PR
- [ ] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [ ] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

This PR is one of 4 changesets included in PR https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3798. My goal here is to merge those changes separately, going in reverse order. This PR is the first (1/4) of the changes.

This PR removes our reliance on the `release` branch, which will in turn make release testing much easier. This is because the old step of merging `master` into `release` made testing PR branches much more difficult.

This PR was tested on my fork:

* Branch: https://github.com/joshuacwnewton/spinalcordtoolbox/tree/jn/3732-release-sct-from-master-branch
* CI run: https://github.com/joshuacwnewton/spinalcordtoolbox/runs/6937561646?check_suite_focus=true
* Resulting release: https://github.com/joshuacwnewton/spinalcordtoolbox/releases/tag/5.7
* Tag with bot commits: https://github.com/joshuacwnewton/spinalcordtoolbox/commits/5.7

## Linked issues

Fixes #3732.

### CODE REVIEW

**Comment by mguaypaq** [comment]
A quick comment before I do an actual review: I think this PR has an unintended side effect, of no longer cleaning up files under dev/ that used to be removed when merging the release branch. I would like to understand this better before approving the PR. Here are a few observations I gathered, before heading home:

- For the 5.6 release, the effect of merging master into the release branch made the dev/ folder reduce in size from 5.4MB to 1.1MB.
- This matches up with the (uncompressed) size difference between the official 5.6 release (7.7MB) and your testing version of a 5.7 release (12MB).

I did a quick count of how many files under each subfolder of dev/ are affected, and here are the results. At a glance, the TL;DR is that some files are removed, some files survive, and I don't see a clear pattern:
```
bef: files present in master
aft: files present in release
del: files removed when merging master into release
(all numbers cumulative in the subtree)

bef aft del
637 100 537 dev/
  4   2   2 dev/GM_atlas/
 83  23  60 dev/atlas/
 11   5   6 dev/atlas/create_atlas/
 16   5  11 dev/atlas/raw_data/
  2   2   0 dev/atlas/raw_data/gm/
 51  10  41 dev/atlas/validate_atlas/
  3   0   3 dev/atlas/validate_atlas/DICE_coefficient/
 32   0  32 dev/atlas/validate_atlas/manual_masks/
  6   0   6 dev/atlas/validate_atlas/manual_masks/benjamin/
  3   0   3 dev/atlas/validate_atlas/manual_masks/charles/
  6   0   6 dev/atlas/validate_atlas/manual_masks/julien/
  6   0   6 dev/atlas/validate_atlas/manual_masks/simon/
  6   0   6 dev/atlas/validate_atlas/manual_masks/tanguy/
 32   2  30 dev/denoise/
 14   0  14 dev/denoise/aonlm/
  6   0   6 dev/denoise/aonlm/wavelet/
 14   1  13 dev/denoise/ornlm/
  6   0   6 dev/denoise/ornlm/wavelet/
  2   2   0 dev/detect_c2c3/
  1   1   0 dev/docs/
  4   3   1 dev/githook/
 12  12   0 dev/install/
  1   1   0 dev/install/old/
 20   0  20 dev/isct_crop_image/
  2   0   2 dev/isct_dice_coefficient/
  3   0   3 dev/isct_orientation3d/
452  38 414 dev/isct_propseg/
400   0 400 dev/isct_propseg/documentation/
399   0 399 dev/isct_propseg/documentation/html/
232   0 232 dev/isct_propseg/documentation/html/search/
 11   5   6 dev/isct_propseg/util/
  2   1   1 dev/itk_transfo/
  1   1   0 dev/scripts/
  3   2   1 dev/spinal_level/
```

**Comment by joshuacwnewton** [comment]
> At a glance, the TL;DR is that some files are removed, some files survive, and I don't see a clear pattern:

I did some sleuthing using [`git rev-list -n 1 HEAD -- [file_path]`](https://stackoverflow.com/a/56054951), and I think I have a timeline that explains the situation:

* **May 5, 2016**: A commit (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/1dd49e5ec16aaf665485f3a04ace145a443b580d) is added to the `release` branch that removes [_the entire `dev/` folder_](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/1dd49e5ec16aaf665485f3a04ace145a443b580d) (1600+ files) from the release branch.
* **2016-2018**: `master` is [repeatedly merged into `release`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/compare/release), gradually repopulating `dev/` a handful of files at a time, whenever `dev/` files on `master` are modified.

* **March 2018**: After 2 years of repopulating `dev/` on `release`, `dev/` ends up in [this state](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/b61a81b05a2c1485db0cdf401b35631b66408f1e/dev), containing 120 total files:

   ```console
   (venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/dev$ find . -type f | wc -l
   120
   ```

* **March 4, 2018**: PR [#1631](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/1631) is merged to `master` (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/0590a7c38cfb31e59d5478c2ad941558a80590fb) and removes _a subset (1000+ files)_ from the `dev/` folder. After this PR is merged, `dev/` ends up in [this state](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/0590a7c38cfb31e59d5478c2ad941558a80590fb/dev), containing 588 files:

   ```
   (venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/dev$ find . -type f | wc -l 
   588
   ```

* **May 27, 2018**: Once again, `master` is merged into `release` (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/43d312ee0b63e293b4f7dc03fbd53a7e88300a58), with the new March deletions carrying over to the existing 120 `dev/` files on `release`. There are conflicts. After resolving said conflicts, `release` contains only 63 files:

   ```
   (venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/dev$ find . -type f | wc -l
   63
   ```

If you take the 588 `dev/` files on master in 2018, and subtract the 63 `dev/` files on `release` in 2018, you get 525, which is roughly the difference observed by yourself, @mguaypaq, when comparing the two branches in 2022 https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3813#issuecomment-1163479681.

So, tl;dr: The discrepancies between `master` and `release` stems from 2 separate clean-up efforts in 2016 and 2018, along with a bunch of merges from `master` into `release` that muddle up the history.

**Comment by mguaypaq** [comment]
It looks like the intent is for `dev/` to exist in the repository for cloning purposes, but not in the release tarballs. In which case, I would suggest adding a step in `create-release.yml` to do `rm -r dev/`, maybe where the merge step used to be. (If you pick this option, it would make sense to include it in this PR.)

Then again, moving `dev/` to its own repo makes a lot of sense. (If you pick this option, maybe you don't need to do it in this PR.)

**Comment by mguaypaq** [comment]
Also, that's some amazing sleuthing you did!

**Comment by joshuacwnewton** [comment]
With #3827 merged, I tested the workflow on my fork using [this CI run](https://github.com/joshuacwnewton/spinalcordtoolbox/actions/runs/2610925657) to produce [this test release](https://github.com/joshuacwnewton/spinalcordtoolbox/releases/tag/5.7.3).

The `dev/` folder is no longer present. The uncompressed sizes that I'm seeing are 9.8MB (before #3827) and 5.8MB (after #3827). 

I think this PR is good to go, now?

**Review by mguaypaq — APPROVED** [review]
Given that #3827 is now dealt with, I'd say LGTM. Thanks for the simpler workflow!
